강원대 FE 유다연 - 3단계 로그인 페이지 UI 만들기

daaoooy opened the pull request:
## [2단계 피드백 사항]
- [X] 의미있는 숫자 상수화
- [X] 구체적인 타입을 위해 as const를 활용
- [X] console.log (선택된 것을 확인하기 위해 추가한 것이라 아직 지우진 않았습니다)
- [X] 스타일 컴포넌트를 import 하는 방식 통일하기

<br />

## [3단계 요구 사항]
- [x] 본인만의 기준으로 일관된 코드를 작성하기
- [x] 기능 단위로 나누어 커밋하기

#### 로그인

- [x] /login 페이지 접속 시 로그인 화면으로 연결되도록 하기 
- [x] 로그인 화면을 만들기
- [x] 로그인 버튼을 클릭하면 이전에 접속한 페이지로 Redirect 되도록 함. 만약 이전 페이지가 없다면 홈(/) 으로 Redirect 되도록 함. (ID, PW 기능은 구현하지 않아도 괜찮아도 됨)
- [x] 네비게이션 바에 로그인과 뒤로가기 버튼을 추가하고 기능 구현하기

#### Not Found 페이지

- [x] 잘못된 경로로 접속 시 Not Found 페이지로 연결되도록 함

#### 선물하기 메인 > 실시간 급상승 선물랭킹

- [x] 실시간 급상승 선물랭킹에서 선택한 필터(성별, 주제)가 새로고침 되어도 유지되도록 변경하기

#### 리팩터링

- [x] 재사용 가능성이 높은 컴포넌트를 찾아 별도의 컴포넌트로 만들기
- [ ] (선택) 스토리북을 통해 재사용 가능한 컴포넌트를 문서화 하기

<br />

## [3단계 피드백 사항]
- [x] Icon 인라인 스타일 작성한 부분 수정
- [x] 경로들 상수로 관리하기
- [x] 모든 페이지에 쓰이고 있는 TopNav를 Layout 컴포넌트를 생성해 활용하기
- [x] 잘못된 컴포넌트 이름과 컴포넌트 역할에 맞게 이름 수정
- [x] 오타 수정
- [x] 버튼 컴포넌트 만들어서 코드 리팩터링
- [x] 유효하지 않은 쿼리 스트링 값이 온 경우 기본값으로 처리

daaoooy committed:
docs(readme) : 3단계 (로그인 페이지 UI) 요구사항 작성

- README 파일에 3단계 요구사항 작성함

- 2단계 피드백 내용 삭제 완료

daaoooy committed:
chore(index) : index.html 타이틀 변경

- 웹사이트 타이들을 변경함

daaoooy committed:
chore(build): react-router-dom 및 react-router 패키지 추가

- react-router 설치

- react-router-dom 설치

daaoooy committed:
remove : 불필요한 파일 삭제

daaoooy committed:
remove : 커밋 실수로 인해 생긴 중복 파일 삭제

daaoooy committed:
chore(asset): not found 이미지, kakao 타이틀 아이콘 추가

daaoooy committed:
refactor : 리팩토링. 폴더 구조 변경 및 파일 이름 변경

- 전체적인 폴더 구조를 변경함

- features 폴더를 생성하여 페이지별로 나누어줌

- 공통으로 사용하는 파일은 src/component 파일에 넣어줌

daaoooy committed:
feat(rank) : 선택한 필터 유지 기능 추가

- 폴더 위치 변경 이루어짐

- 실시간 급상승 선물 랭킹에서 선택한 필터 (성별, 주제)가 새로고침이 되어도 유지되도록 변경

daaoooy committed:
feat(top-nav) : 네비게이션 바에 기능 추가

- 로그인과 뒤로가기 버튼을 추가하고 기능 구현

- 프로필 아이콘 클릭 시, 로그인 화면으로 / 뒤로가기 아이콘 클릭시, 이전 화면으로 이동

- 폴더 위치 변경이 이루어짐 (공통으로 사용하는 컴포넌트 폴더에 위치)

daaoooy committed:
feat(login) : 로그인 페이지 추가 및 기능 구현

- 로그인 페이지 구현

- 로그인 버튼 클릭시 이전 페이지로, 이전페이지가 없다면 홈으로 이동

daaoooy committed:
feat(product) : 상품 아이템 컴포넌트 분리

- 재사용을 위해 TrendingGiftRanking 에 있던 프로덕트 컴포넌트를 분리함

daaoooy committed:
feat(page) : 라우팅 구현을 위해 페이지 별로 파일 생성 및 구현

daaoooy committed:
feat(not-found) : Not Found 페이지 추가 및 기능 구현

- 잘못된 경로로 이동시 나타낼 페이지 생성

daaoooy committed:
feat(router) : 라우터 기능 구현

daaoooy committed:
refactor(global) : 배경색 수정

daaoooy committed:
refactor : 커밋 실수로 생긴 중복된 파일 삭제

Leejin-Yang commented on index.html:
```diff
@@ -8,7 +8,7 @@
       rel="stylesheet"
     />
-    <title>Vite + React + TS</title>
+    <title>강원대 FE 유다연</title>
```
🙌

Leejin-Yang commented on src/component/TopNavigationBar/TopNavigationBar.tsx:
```diff
@@ -0,0 +1,36 @@
+
+  return (
+    <Nav>
+      <Icon onClick={handleGoBack} style={{ cursor: 'pointer' }}>
```
인라인 스타일로 작성한 부분이 있네요, Icon 스타일 컴포넌트에 추가할 수 있지 않을까요?

daaoooy replied:
> 인라인 스타일로 작성한 부분이 있네요, Icon 스타일 컴포넌트에 추가할 수 있지 않을까요?

수정 완료

Leejin-Yang commented on src/component/TopNavigationBar/TopNavigationBar.tsx:
```diff
@@ -0,0 +1,36 @@
+
+  const handleGoBack = () => {
+    if (window.history.length > 1) {
+      navigate(-1)
```
외부 페이지에서 앱으로 들어왔을 때 뒤로가기 버튼을 누른다면 어떻게 될까요?

daaoooy replied:
> 외부 페이지에서 앱으로 들어왔을 때 뒤로가기 버튼을 누른다면 어떻게 될까요?

테스트로 구글 페이지 -> localhost:5173 으로 이동해보고, 구현한 뒤로가기 버튼을 누르니까 구글 페이지로 옵니다! 혹시 이렇게 되면 문제가 있는 걸까요!

Leejin-Yang replied:
앱 내부에 있는 뒤로가기 버튼이라면, 앱 내에서 정의한 이동 경로로 돌아가는 동작을 기대할 수 있어요 😊

Leejin-Yang commented on src/routes/Router.tsx:
```diff
@@ -0,0 +1,17 @@
+const AppRouter = () => {
+  return (
+    <Routes>
+      <Route path="/" element={<GiftPage />} />
+      <Route path="/login" element={<LoginPage />} />
+      <Route path="*" element={<NotFoundPage />} />
```
경로들도 상수로 관리하면 유지보수가 더 쉬울 거에요 😊

daaoooy replied:
> 경로들도 상수로 관리하면 유지보수가 더 쉬울 거에요 😊

수정 완료

Leejin-Yang commented on src/features/NotFound/pages/NotFoundPage.tsx:
```diff
@@ -0,0 +1,13 @@
+const NotFoundPage = () => {
+  return (
+    <>
+      <TopNavigationBar />
```
지금 이 컴포넌트가 모든 페이지에서 쓰이고 있는데, 중복을 줄일 수 있지 않을까요? Layout 컴포넌트를 생성해 활용할 수 있어보여요 😊

daaoooy replied:
> 지금 이 컴포넌트가 모든 페이지에서 쓰이고 있는데, 중복을 줄일 수 있지 않을까요? Layout 컴포넌트를 생성해 활용할 수 있어보여요 😊

수정 완료

Leejin-Yang commented on src/features/Login/pages/LoginPage.tsx:
```diff
@@ -0,0 +1,13 @@
+import TopNavigationBar from '@/component/TopNavigationBar/TopNavigationBar'
+import Input from '@/features/Login/components/Input'
+
+const GiftPage = () => {
```
파일대로라면 LoginPage가 맞아보이네요..!

daaoooy replied:
> 파일대로라면 LoginPage가 맞아보이네요..!

수정 완료

Leejin-Yang commented on src/features/Login/components/Input.tsx:
```diff
@@ -0,0 +1,38 @@
+import { useNavigate, useLocation } from 'react-router-dom'
+import KaKaoTitleIcon from '@/assets/icons/kakao-title.svg?react'
+
+const Input: React.FC = () => {
```
컴포넌트의 역할을 봤을 때, Input 보단 LoginForm과 같은 이름이 어울려보여요. 보통 Input은 하나의 input을 나타낼 수 있기 때문에 컴포넌트 역할에 맞게 수정하는게 좋아보여요 😊

daaoooy replied:
> 컴포넌트의 역할을 봤을 때, Input 보단 LoginForm과 같은 이름이 어울려보여요. 보통 Input은 하나의 input을 나타낼 수 있기 때문에 컴포넌트 역할에 맞게 수정하는게 좋아보여요 😊

수정 완료

Leejin-Yang commented on src/features/Login/components/Input.tsx:
```diff
@@ -0,0 +1,38 @@
+        <KakaoTitle>
+          <KaKaoTitleIcon />
+        </KakaoTitle>
+        <InputForm placeholder="이메일" type="emial"></InputForm>
+        <InputForm placeholder="비밀번호" type="password"></InputForm>
```
이메일 오타가 있네요! 그리고 아래와 같이 수정할 수 있어보여요 😊

```suggestion
        <InputForm placeholder="이메일" type="email" />
        <InputForm placeholder="비밀번호" type="password" />
```

daaoooy replied:
> 이메일 오타가 있네요! 그리고 아래와 같이 수정할 수 있어보여요 😊

수정 완료

Leejin-Yang commented on README.md:
```diff
@@ -49,62 +43,39 @@
+##### 리팩터링
 
-👧 : [수정 완료] theme 폴더를 생성하여 각각 분리해주었습니다. 이런식으로 하면 괜찮을까요?
+- [x] 재사용 가능성이 높은 컴포넌트를 찾아 별도의 컴포넌트로 만들기
```
이 부분에 대해 한가지 시도해볼 수 있는걸 추천하자면,
Button 컴포넌트를 만들어보는건 어떨까요? 접기, 로그인, 홈으로 버튼이 비슷한 스타일을 가지고 있어서 한번 시도할 수 있어보여요!

많은 것을 추가할 건 없고, 현재 단계에서 보이는 정도의 variation으로 간단하게 한번 리팩터링 해보면 될 것 같아요.

한가지 팁을 드리자면 기본 html 속성은 리액트에서 타입을 지원하는데요, Props 타입을 선언할 때 [ComponentPropsWithoutRef](https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#wrappingmirroring-a-html-element)를 사용하면 onClick과 같은 기본 속성의 타입을 지원해줄 거에요 😊

daaoooy replied:
> 이 부분에 대해 한가지 시도해볼 수 있는걸 추천하자면, Button 컴포넌트를 만들어보는건 어떨까요? 접기, 로그인, 홈으로 버튼이 비슷한 스타일을 가지고 있어서 한번 시도할 수 있어보여요!
> 
> 많은 것을 추가할 건 없고, 현재 단계에서 보이는 정도의 variation으로 간단하게 한번 리팩터링 해보면 될 것 같아요.
> 
> 한가지 팁을 드리자면 기본 html 속성은 리액트에서 타입을 지원하는데요, Props 타입을 선언할 때 [ComponentPropsWithoutRef](https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#wrappingmirroring-a-html-element)를 사용하면 onClick과 같은 기본 속성의 타입을 지원해줄 거에요 😊

-> 수정해보려고 했는데, 음.. 이런식으로 리팩터링 하는 것이 맞는지 잘 모르겠습니다.. 일단 수정해서 PR 하겠습니다.

Leejin-Yang replied:
추천으로 제안 드린 부분이라, 아직 확신이 없거나 부담이 된다면 지금 단계에서 적용하지 않아도 괜찮아요. 진행하다 어려운 점이 있으면 언제든 질문 남겨주세요!

Leejin-Yang commented on src/features/Gift/components/TrendingGiftRanking/TrendingGiftRanking.tsx:
```diff
@@ -1,34 +1,52 @@
-  const [selectedType, setSelectedType] = useState<TypeLabel | ''>('')
+  const [searchParams, setSearchParams] = useSearchParams()
+
+  const selectedGender = searchParams.get('gender') ?? genderList[0].label
+  const selectedType = searchParams.get('type') ?? typeList[0]
```
기본값을 잘 정해주셨네요 👍
다만, 유효하지 않은 쿼리 스트링 값이 온 경우 아래와 같이 필터가 활성화 되어있지 않네요. 유효하지 않은 경우 url과 필터를 기본값으로 가도록 해보는건 어떨까요?

<img width="1060" alt="Screenshot 2025-07-01 at 02 49 30" src="https://github.com/user-attachments/assets/c1a0ea91-85f4-4c23-9b29-518e3bbbce4e" />

daaoooy replied:
> 기본값을 잘 정해주셨네요 👍 다만, 유효하지 않은 쿼리 스트링 값이 온 경우 아래와 같이 필터가 활성화 되어있지 않네요. 유효하지 않은 경우 url과 필터를 기본값으로 가도록 해보는건 어떨까요?
> 
> <img alt="Screenshot 2025-07-01 at 02 49 30" width="1060" src="https://private-user-images.githubusercontent.com/78616893/460808399-c1a0ea91-85f4-4c23-9b29-518e3bbbce4e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTEzMzU2MTAsIm5iZiI6MTc1MTMzNTMxMCwicGF0aCI6Ii83ODYxNjg5My80NjA4MDgzOTktYzFhMGVhOTEtODVmNC00YzIzLTliMjktNTE4ZTNiYmJjZTRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAxVDAyMDE1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxMzljZGYwNjY5OGFlMGJkNGQ4YTIzOTVjNGMyNTA0MjA5ZjIyOWYzN2ExZDRmNGM1MzFiZGZhN2RkZDljMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BGUZCMxtF6f246LvPTMOATFGFRwtU87Lp1bLgYajIGY">

-> useEffect와 타당성을 확인하기 위한 isValid 변수를 통해 수정해보았습니다. 마저 수정 후 PR 하겠습니다.

Leejin-Yang reviewed (changes requested):
다연님 안녕하세요 👋

이번 단계도 잘 진행해주셨네요.! 지난 피드백도 잘 반영해주셔서 감사합니다 😊
전반적으로 요구사항도 잘 지켜주시고, UI도 깔끔하게 구현했다는 생각이 듭니다 👍👍
몇가지 소소한 코멘트 남겼는데, 수정한 부분을 같이 보고싶어서 Request Changes로 남겨요.

리뷰 확인해주시고 요청 보내주세요. 궁금한 점은 언제든지 코멘트나 DM으로 남겨주세요 🙌

daaoooy committed:
refactor(button) : Button  컴포넌트 생성 및 피드백 수정

- 버튼 컴포넌트를 만들어 코드 리팩터링 진행

- Input -> LoginForm 으로 파일명 변경

- email 오타 수정

daaoooy committed:
refactor :인라인 스타일 수정

daaoooy committed:
feat(filter) : 유효하지 않은 쿼리 파라미터 처리 기능 추가

daaoooy committed:
feat : 경로 상수 설정 및 NavLayout 컴포넌트 생성

- 라우팅에 사용할 경로를 상수로 설정함

- 공통적으로 사용되는 TopNavigationBar 을 처리하기 위해 NavLayout을 생성하고 사용함

Leejin-Yang commented on src/features/Gift/components/TrendingGiftRanking/TrendingGiftRanking.tsx:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  // 피드백 : 유효하지 않은 쿼리 스트링의 경우 처리
+  useEffect(() => {
+    const params = new URLSearchParams(searchParams)
+    let isValid = true
+
+    if (!genderList.map((g) => g.label).includes(selectedGender as Gender)) {
+      params.set('gender', genderList[0].label)
+      isValid = false
+    }
+
+    if (!typeList.includes(selectedType as Type)) {
+      params.set('type', typeList[0])
+      isValid = false
+    }
+
+    if (!isValid) setSearchParams(params, { replace: true })
+  }, [searchParams, selectedGender, selectedType, setSearchParams])
```
잘 작성해주셨네요 👍
다만, 위의 핸들러 함수를 통해 searchParams를 업데이트했을 때 이 이펙트도 다시 실행되면서 중복으로 동작할 수 있으니, 이 부분은 수정해보면 좋아보입니다 😊

Leejin-Yang commented on src/component/Layout/NavLayout.tsx:
```diff
@@ -0,0 +1,13 @@
+import { Outlet } from 'react-router-dom'
+import TopNavigationBar from '@/component/TopNavigationBar/TopNavigationBar'
+
+const Layout = () => {
+  return (
+    <>
+      <TopNavigationBar />
+      <Outlet />
+    </>
+  )
+}
```
👍

Leejin-Yang commented on src/component/Button/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+  },
+}
+
+const StyledButton = styled.button<{ $buttonType: ButtonType }>`
```
버튼 컴포넌트도 잘 작성해주셨네요 👍

지금은 각 용도에 맞게 타입을 받도록 했지만, 디자인적 관점에서 variant를 나눠보는 것도 좋아보여요. 버튼의 사이즈, 배경 유무 등 같은 경우로요. 이렇게 하면 전체 프로젝트에 디자인적으로 통일성도 가져올 수 있어요. 지금 단계에서는 크게 고민할 부분은 아니고 이후 UI를 더 작성하면서 이 버튼을 활용할 때 어떻게 속성들을 추가해야할 지 고민해보시면 될 것 같아요 😊

Leejin-Yang reviewed (approved):
다연님 안녕하세요 👋

이번 미션 끝까지 잘 진행해주셔서 감사합니다 👍
그리고 피드백도 확인해주시고 잘 반영해주셔서 감사해요 ☺️

이번 미션 리뷰는 여기까지지만, 학습하면서 궁금한 점은 언제든지 디스코드 DM 보내주세요 😊
다음 미션도 응원하겠습니다! 고생 많으셨어요 🙌
